File: 2024/1/main.py

import collections

if __name__ == "__main__":
    numbers = [[],[]]
    with open("data") as f:
        for line in (ln.strip().split() for ln in f.readlines()):
            for i,n in enumerate(line):
                numbers[i].append(int(n))
    
    numbers[0].sort()
    numbers[1].sort()
    dists = 0
    seen = collections.defaultdict(int)
    for i,_ in enumerate(numbers[0]):
        seen[numbers[1][i]] += 1 
        dists += abs(numbers[0][i] - numbers[1][i])
    print(f"1: {dists}")

    msum = 0
    for n in numbers[0]:
        msum += n*seen[n]
    print(f"2: {msum}")

File: 2024/1/main.go
package main

import (
	"os"
	"strings"
	"strconv"
	"slices"
	"fmt"
)

func main(){
	dat,_ := os.ReadFile("data")
	lines := strings.Split(string(dat),"\n")
	n := len(lines)
	lefts := make([]int,n)
	rights := make([]int,n)
	for i,line := range(lines) {
		if line != "" {
			numbers := strings.Split(line,"   ")
			lefts[i],_ = strconv.Atoi(numbers[0])
			rights[i],_ = strconv.Atoi(numbers[1])
		}
	}
	slices.Sort(lefts)
	slices.Sort(rights)
	dists := 0
	seens := make(map[int]int)
	for i:=0;i<n;i++ {
		x,y := lefts[i],rights[i]
		if x > y {
			dists += x - y
		} else {
			dists += y - x
		}
		_,success := seens[y]
		if success {
			seens[y]++
		} else {
			seens[y] = 1
		}
	}
	msum := 0
	for i:=0;i<n;i++ {
		x := lefts[i]
		seen,success := seens[x]
		if !success{
			seen = 0
		}
		msum += lefts[i]*seen
	}
	fmt.Printf("1: %v\n",dists)
	fmt.Printf("2: %v\n",msum)
}
